File: mobile/lib/domain/entities/booking.dart
import 'package:equatable/equatable.dart';

enum BookingStatus { pending, confirmed, inProgress, completed, cancelled }

class Booking extends Equatable {
  const Booking({
    required this.id,
    required this.serviceId,
    required this.serviceName,
    required this.userId,
    required this.status,
    required this.scheduledAt,
    required this.address,
    required this.totalAmount,
    this.notes,
    this.providerId,
    this.providerName,
    this.createdAt,
  });

  final String id;
  final String serviceId;
  final String serviceName;
  final String userId;
  final BookingStatus status;
  final DateTime scheduledAt;
  final String address;
  final double totalAmount;
  final String? notes;
  final String? providerId;
  final String? providerName;
  final DateTime? createdAt;

  bool get isActive =>
      status == BookingStatus.pending || status == BookingStatus.confirmed;

  bool get isCancellable => isActive;

  @override
  List<Object?> get props => [
        id, serviceId, serviceName, userId, status,
        scheduledAt, address, totalAmount, notes,
        providerId, providerName, createdAt,
      ];
}

File: backend/app/models/booking.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Booking(Base):
    __tablename__ = "bookings"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    service_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("services.id", ondelete="RESTRICT"), nullable=False
    )
    # Status values: PENDING, CONFIRMED, COMPLETED, CANCELLED
    status: Mapped[str] = mapped_column(String(20), default="PENDING", nullable=False)
    scheduled_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    user: Mapped["User"] = relationship("User", back_populates="bookings")
    service: Mapped["Service"] = relationship("Service", back_populates="bookings")
    payment: Mapped["Payment"] = relationship(
        "Payment", back_populates="booking", uselist=False
    )
